pkg/client/client.go

```diff
@@ -557,6 +562,177 @@ func (c *Client) Reset(ctx context.Context, state string) error {
+}
+
+func cleanupCacheLockFile(lockFile *os.File) error {
+	err := os.Remove(lockFile.Name())
+	if err != nil {
+		return fmt.Errorf("unable to remove cache lock file: %w", err)
+	}
+
+	return nil
+}
+
+func (c *Client) GetCache(ctx context.Context, cacheRootDir string) (int64, error) {
+	objectDir := CacheObjectsDir(cacheRootDir)
+	err := os.MkdirAll(objectDir, 0755)
+	if err != nil {
+		return -1, fmt.Errorf("cannot create object folder: %w", err)
+	}
+
+	tmpObjectDir := CacheTmpDir(cacheRootDir)
+	os.RemoveAll(tmpObjectDir)
+	err = os.MkdirAll(tmpObjectDir, 0755)
+	if err != nil {
+		return -1, fmt.Errorf("cannot create tmp folder to unpack cached objects: %w", err)
+	}
+
+	lockFile, err := obtainCacheLockFile(cacheRootDir)
+	if err != nil {
+		return -1, err
+	}
+
+	availableVersions := ReadCacheVersionFile(cacheRootDir)
+
+	ctx, span := telemetry.Start(ctx, "client.get_cache")
+	defer span.End()
+
+	request := &pb.GetCacheRequest{}
+
+	stream, err := c.fs.GetCache(ctx, request)
+	if err != nil {
+		return -1, fmt.Errorf("fs.GetCache connect: %w", err)
+	}
+
+	// Pull one response before booting workers
+	// This is a short circuit in case the cache doesn't exist
+	response, err := stream.Recv()
+	if err == io.EOF {
+		err = cleanupCacheLockFile(lockFile)
+		if err != nil {
+			return -1, err
+		}
+		os.RemoveAll(tmpObjectDir)
+		return -1, nil
+	}
+	if err != nil {
+		return -1, fmt.Errorf("fs.GetCache receive: %w", err)
+	}
+	version := response.Version
+
+	for _, availableVersion := range availableVersions {
+		if version == availableVersion {
+			return availableVersions[len(availableVersions)-1], nil
+		}
+	}
+
+	tarChan := make(chan *pb.GetCacheResponse, 16)
+	group, ctx := errgroup.WithContext(ctx)
+	ctx, cancel := context.WithCancel(ctx)
+
+	group.Go(func() error {
+		ctx, span := telemetry.Start(ctx, "cache-object-receiver")
+		defer span.End()
+		defer close(tarChan)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return nil
+			case tarChan <- response:
+				response, err = stream.Recv()
+				if err == io.EOF {
+					return nil
+				}
+				if response.Version != version {
+					cancel()
+					return fmt.Errorf("fs.GetCache version mismatch: %d, %d", response.Version, version)
+				}
+				if err != nil {
+					cancel()
+					return fmt.Errorf("fs.GetCache receive: %w", err)
+				}
+			}
+		}
+	})
+
+	for i := 0; i < parallelWorkerCount(); i++ {
+		// create the attribute here when `i` is different
+		attr := key.Worker.Attribute(i)
+
+		group.Go(func() error {
+			ctx, span := telemetry.Start(ctx, "cache-object-writer", trace.WithAttributes(attr))
+			defer span.End()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case response, ok := <-tarChan:
+					if !ok {
+						return nil
+					}
+
+					tarReader := db.NewTarReader(response.Bytes)
+					hashHex := hex.EncodeToString(response.Hash)
+					tempDest := filepath.Join(tmpObjectDir, hashHex)
+					finalDest := filepath.Join(objectDir, hashHex)
```

Comment:
Well, the thing is that we don't know that the extraction will succeed. If we extract directly to the final destination we risk creating an invalid file structure. At present, even if we wrote a bad file, we would not use it, but I can't guarantee that this will hold, and I haven't written tests to prove it either.

The easiest way to avoid that situation is to extract somewhere else and avoid moving if the extraction fails.

Reply:
Ahh got it.

Our two options are to work in a temp directory (as you are doing) or do a full cleanup (`rm -rf`) at the beginning of Cache Creation to potentially drop any failed previous attempt.

Is that second option possible / simpler, or not something we can do in this case.

Reply:
A worker could have an in-flight request using existing cache versions. We can *never* delete what has been made available, unless we implement synchronization

Reply:
Yeah 100%, we should never delete something once it has been made available.

But if it was never made available then it should be safe to delete?

Reply:
yes, if and only if it was never made available, we can delete it, the trick is in knowing